File: 02_first/linear_regression.py
import numpy as np


def predict(X, w):
    return X * w


def loss(X, Y, w):
    return np.average((predict(X, w) - Y) ** 2)


def train(X, Y, iterations, lr):
    w = 0
    for i in range(iterations):
        current_loss = loss(X, Y, w)
        print(f"Iteration {i:4d} => Loss: {current_loss:.6f}")

        if loss(X, Y, w + lr) < current_loss:
            w += lr
        elif loss(X, Y, w - lr) < current_loss:
            w -= lr
        else:
            return w

    raise Exception("Couldn't converge within %d iterations" % iterations)


# Import the dataset
X, Y = np.loadtxt("pizza.txt", skiprows=1, unpack=True)

# Train the system
w = train(X, Y, iterations=10000, lr=0.01)
print(f"\nw={w:.3f}")

# Predict the number of pizzas
x = 20
print(f"Prediction: x={x:d} => y={predict(x, w):.2f}")

# Plot the chart
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_theme()
plt.plot(X, Y, "bo")
plt.xticks(fontsize=15)
plt.yticks(fontsize=15)
plt.xlabel("Reservations", fontsize=30)
plt.ylabel("Pizzas", fontsize=30)
x_edge, y_edge = 50, 50
plt.axis([0, x_edge, 0, y_edge])
plt.plot([0, x_edge], [0, predict(x_edge, w)], linewidth=1.0, color="g")
plt.show()

File: 02_first/linear_regression.jl
using Statistics
using Printf
using Plots
using Plots.PlotMeasures

function predict(X, w)
   return X * w
end

function loss(X, Y, w)
    tmp = predict(X, w) - Y
    return Statistics.mean(tmp .^2)
end

function train(X, Y; iterations, lr)
    w = 0.0
    for i = 0:iterations
        current_loss = loss(X, Y, w)
        @printf("Iteration %4d => Loss: %.6f\n", i, current_loss)
        if loss(X, Y, w + lr) < current_loss
            w += lr
        elseif loss(X, Y, w - lr) < current_loss
            w -= lr
        else
            return w
        end
    end
end

# Import the dataset
data = readdlm("pizza.txt", skipstart=1)
X = data[:, 1]
Y = data[:, 2]

# Train the system
w = train(X, Y, iterations=10000, lr=0.01)
 @printf("w = %.3f\n", w)

# Predict the number of pizzas
@printf("Prediction: x = %d => y = %.2f\n", 20, predict(20, w))

# Calculate the least-square solution
A = [ones(length(X)) X]
b = Y
x = A \ b
@printf("LS: w = %.3f\n", x[2])
@printf("LS: b = %.3f\n", x[1])
x0 = 20
@printf("Prediction: x = %d => y = %.2f\n", x0, x[1]+ x0*x[2])

# Plot the chart
x_edge = 30
y_edge = 30
x = range(0, x_edge, length=5)
y = predict.(x, w)
p = plot(x, y, color=:green, label="prediction")
scatter!(X, Y, color=:blue, label="observations")

plot!(p, margin=5mm,
    tickfontsize=15,
    xlabel="Reservations",
    ylabel="Pizzas",
    guidefontsize=20,
)

